apis/apps/v1alpha1/containerrecreaterequest_types.go

```diff
@@ -0,0 +1,182 @@
+type ContainerRecreateRequestSpec struct {
+	// Containers contains the containers that need to recreate in the Pod.
+	Containers []ContainerRecreateRequestContainer `json:"containers"`
+	// Strategy defines strategies for containers recreation.
+	Strategy *ContainerRecreateRequestStrategy `json:"strategy,omitempty"`
+	// ActiveDeadlineSeconds is the deadline duration of this ContainerRecreateRequest.
+	ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty"`
+	// TTLSecondsAfterFinished is the TTL duration after this ContainerRecreateRequest has completed.
+	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+}
+
+// ContainerRecreateRequestContainer defines the container that need to recreate.
+type ContainerRecreateRequestContainer struct {
+	// Name of the container that need to recreate.
+	// It must be existing in the real Pod containers.
+	Name string `json:"name"`
+	// SkipPreStop indicates whether to skip the preStop hook during recreating this container.
+	SkipPreStop bool `json:"skipPreStop,omitempty"`
+	// StatusContext is synced from the real container in Pod spec during this ContainerRecreateRequest creating.
+	// Read-only.
+	PreStop *v1.Handler `json:"preStop,omitempty"`
+	// StatusContext is synced from the real container in Pod spec during this ContainerRecreateRequest creating.
+	// Read-only.
+	Ports []v1.ContainerPort `json:"ports,omitempty"`
+	// StatusContext is synced from the real Pod status during this ContainerRecreateRequest creating.
+	// Read-only.
+	StatusContext *ContainerRecreateRequestContainerContext `json:"statusContext,omitempty"`
+}
+
+// ContainerRecreateRequestContainerContext contains context status of the container that need to recreate.
+type ContainerRecreateRequestContainerContext struct {
+	// Container's ID in the format 'docker://<container_id>'.
+	ContainerID string `json:"containerID"`
+	// The number of times the container has been restarted, currently based on
+	// the number of dead containers that have not yet been removed.
+	// Note that this is calculated from dead containers. But those containers are subject to
+	// garbage collection. This value will get capped at 5 by GC.
+	RestartCount int32 `json:"restartCount"`
+}
+
+// ContainerRecreateRequestStrategy contains the strategies for containers recreation.
+type ContainerRecreateRequestStrategy struct {
+	// FailurePolicy decides whether to continue if one container fails to recreate
+	FailurePolicy ContainerRecreateRequestFailurePolicyType `json:"failurePolicy,omitempty"`
+	// The priority of grace period usage:
+	// stopGracePeriodSeconds > pod.Spec.TerminationGracePeriodSeconds > 3s
+	StopGracePeriodSeconds *int64 `json:"stopGracePeriodSeconds,omitempty"`
+	// ContinueByReady indicates whether to recreate the next container only if the previous one has been recreated and ready
+	ContinueByReady bool `json:"continueByReady,omitempty"`
```

Comment:
please consider release ContinueByReady and ReadyProtection features in future patches